mlir/lib/Analysis/PresburgerSet.cpp

```diff
@@ -318,6 +318,52 @@ bool PresburgerSet::findIntegerSample(SmallVectorImpl<int64_t> &sample) {
   return false;
 }
 
+PresburgerSet PresburgerSet::coalesce() {
+  assert(nSym == 0 && "symbolic sets are not yet supported by coalesce");
+  PresburgerSet newSet =
+      PresburgerSet::getEmptySet(this->getNumDims(), this->getNumSyms());
+  ArrayRef<FlatAffineConstraints> basicSetVector =
+      flatAffineConstraints;
+  SmallVector<bool, 4> marked(this->getNumFACs());
+
+  for (unsigned i = 0; i < basicSetVector.size(); i++) {
+    if (marked[i])
+      continue;
+    FlatAffineConstraints bs1 = basicSetVector[i];
+
+    unsigned numIneq = bs1.getNumInequalities();
+    unsigned numEq = bs1.getNumEqualities();
+    // check if bs1 is contained in any basicSet
+    for (unsigned j = 0; j < basicSetVector.size(); j++) {
+      if (j == i || marked[j])
+        continue;
+      FlatAffineConstraints bs2 = basicSetVector[j];
+      Simplex simplex(bs2);
+
+      bool contained = true;
```

Comment:
I think you can refactor this out into a separate function, probably `Simplex::contains(const FlatAffineConstraints &fac)`. You can then also return early when any constraint turns out to be non-redundant. (I'm suggesting making it a member function of Simplex rather than FlatAffineConstraints so that you can re-use the same Simplex object many times as suggested above.)

Reply:
Done.
I decided to name the function `Simplex::containedIn(const FlatAffineConstraints &fac)` as it actually checks whether the polytope behind the simplex is `contained` in the specified FAC.